File: libs/sdk-typescript/src/Disk.ts
/*
 * Copyright 2025 Daytona Platforms Inc.
 * SPDX-License-Identifier: Apache-2.0
 */

import { DiskDto, DisksApi } from '@daytonaio/api-client'
import { DaytonaNotFoundError } from './errors/DaytonaError'

/**
 * Represents a Daytona Disk which is persistent storage for Sandboxes.
 *
 * @property {string} id - Unique identifier for the Disk
 * @property {string} name - Name of the Disk
 * @property {string} organizationId - Organization ID that owns the Disk
 * @property {number} size - Disk size in GB
 * @property {string} state - Current state of the Disk
 * @property {string} runnerId - Runner ID where the Disk is located
 * @property {string} errorReason - Error reason if Disk is in error state
 * @property {string} createdAt - Date and time when the Disk was created
 * @property {string} updatedAt - Date and time when the Disk was last updated
 */
export type Disk = DiskDto & { __brand: 'Disk' }

/**
 * Service for managing Daytona Disks.
 *
 * This service provides methods to list, get, create, and delete Disks.
 *
 * @class
 */
export class DiskService {
  constructor(private disksApi: DisksApi) {}

  /**
   * Lists all available Disks.
   *
   * @returns {Promise<Disk[]>} List of all Disks accessible to the user
   *
   * @example
   * const daytona = new Daytona();
   * const disks = await daytona.disk.list();
   * console.log(`Found ${disks.length} disks`);
   * disks.forEach(disk => console.log(`${disk.name} (${disk.id}) - ${disk.size}GB`));
   */
  async list(): Promise<Disk[]> {
    const response = await this.disksApi.listDisks()
    return response.data as Disk[]
  }

  /**
   * Gets a Disk by its ID.
   *
   * @param {string} diskId - ID of the Disk to retrieve
   * @returns {Promise<Disk>} The requested Disk
   * @throws {Error} If the Disk does not exist or cannot be accessed
   *
   * @example
   * const daytona = new Daytona();
   * const disk = await daytona.disk.get("disk-id");
   * console.log(`Disk ${disk.name} is in state ${disk.state}`);
   */
  async get(diskId: string): Promise<Disk> {
    const response = await this.disksApi.getDisk(diskId)
    return response.data as Disk
  }

  /**
   * Creates a new Disk with the specified name and size.
   *
   * @param {string} name - Name for the new Disk
   * @param {number} size - Size of the Disk in GB
   * @returns {Promise<Disk>} The newly created Disk
   * @throws {Error} If the Disk cannot be created
   *
   * @example
   * const daytona = new Daytona();
   * const disk = await daytona.disk.create("my-data-disk", 50);
   * console.log(`Created disk ${disk.name} with ID ${disk.id} (${disk.size}GB)`);
   */
  async create(name: string, size: number): Promise<Disk> {
    const response = await this.disksApi.createDisk({ name, size })
    return response.data as Disk
  }

  /**
   * Deletes a Disk.
   *
   * @param {Disk} disk - Disk to delete
   * @returns {Promise<void>}
   * @throws {Error} If the Disk does not exist or cannot be deleted
   *
   * @example
   * const daytona = new Daytona();
   * const disk = await daytona.disk.get("disk-name");
   * await daytona.disk.delete(disk);
   * console.log("Disk deleted successfully");
   */
  async delete(disk: Disk): Promise<void> {
    await this.disksApi.deleteDisk(disk.id)
  }
}

File: libs/sdk-python/src/daytona/_async/disk.py
# Copyright 2025 Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

from typing import List

from daytona_api_client_async import CreateDiskDto, DisksApi

from ..common.disk import Disk


class AsyncDiskService:
    """Service for managing Daytona Disks. Can be used to list, get, create and delete Disks."""

    def __init__(self, disks_api: DisksApi):
        self.__disks_api = disks_api

    async def list(self) -> List[Disk]:
        """List all Disks.

        Returns:
            List[Disk]: List of all Disks.

        Example:
            ```python
            daytona = AsyncDaytona()
            disks = await daytona.disk.list()
            for disk in disks:
                print(f"{disk.name} ({disk.id}) - {disk.size}GB")
            ```
        """
        disks = await self.__disks_api.list_disks()
        return [Disk.from_dto(disk) for disk in disks]

    async def get(self, disk_id: str) -> Disk:
        """Get a Disk by ID.

        Args:
            disk_id (str): ID of the Disk to get.

        Returns:
            Disk: The Disk object.

        Example:
            ```python
            daytona = AsyncDaytona()
            disk = await daytona.disk.get("disk-id")
            print(f"{disk.name} ({disk.id}) - {disk.size}GB")
            ```
        """
        disk_dto = await self.__disks_api.get_disk(disk_id)
        return Disk.from_dto(disk_dto)

    async def create(self, name: str, size: int) -> Disk:
        """Create a new Disk.

        Args:
            name (str): Name of the Disk to create.
            size (int): Size of the Disk in GB.

        Returns:
            Disk: The Disk object.

        Example:
            ```python
            daytona = AsyncDaytona()
            disk = await daytona.disk.create("test-disk", 50)
            print(f"{disk.name} ({disk.id}); state: {disk.state}; size: {disk.size}GB")
            ```
        """
        disk_dto = await self.__disks_api.create_disk(CreateDiskDto(name=name, size=size))
        return Disk.from_dto(disk_dto)

    async def delete(self, disk: Disk) -> None:
        """Delete a Disk.

        Args:
            disk (Disk): Disk to delete.

        Example:
            ```python
            daytona = AsyncDaytona()
            disk = await daytona.disk.get("test-disk")
            await daytona.disk.delete(disk)
            print("Disk deleted")
            ```
        """
        await self.__disks_api.delete_disk(disk.id)
